File: services/auth-service.ts
import apiClient from "./api-client"
import type { User } from "@/types/user"

export interface LoginRequest {
  username: string
  password: string
}

export interface RegisterRequest {
  username: string
  email: string
  password: string
  fullName: string
}

export interface AuthResponse {
  token: string
  refreshToken: string
  user?: User
}

export interface ForgotPasswordRequest {
  email: string
}

export interface ResetPasswordRequest {
  token: string
  newPassword: string
}

export interface ChangePasswordRequest {
  currentPassword: string
  newPassword: string
}

class AuthService {
  // Login user
  async login(credentials: LoginRequest): Promise<User> {
    try {
      const response = await apiClient.post<AuthResponse>("/auth/login", credentials)
      this.setSession(response.token, response.refreshToken)

      // Fetch user profile if not included in response
      const user = response.user || (await this.getCurrentUser())
      localStorage.setItem("user", JSON.stringify(user))

      return user
    } catch (error) {
      throw error
    }
  }

  // Register new user
  async register(userData: RegisterRequest): Promise<User> {
    try {
      const response = await apiClient.post<AuthResponse>("/auth/register", userData)
      this.setSession(response.token, response.refreshToken)

      // Fetch user profile if not included in response
      const user = response.user || (await this.getCurrentUser())
      localStorage.setItem("user", JSON.stringify(user))

      return user
    } catch (error) {
      throw error
    }
  }

  // Logout user
  async logout(): Promise<void> {
    try {
      const refreshToken = localStorage.getItem("refreshToken")
      if (refreshToken) {
        await apiClient.post("/auth/logout", { refreshToken })
      }
    } catch (error) {
      console.error("Logout error:", error)
    } finally {
      this.clearSession()
    }
  }

  // Get current user profile
  async getCurrentUser(): Promise<User> {
    try {
      return await apiClient.get<User>("/users/me")
    } catch (error) {
      throw error
    }
  }

  // Request password reset
  async forgotPassword(request: ForgotPasswordRequest): Promise<{ message: string }> {
    try {
      return await apiClient.post<{ message: string }>("/auth/forgot-password", request)
    } catch (error) {
      throw error
    }
  }

  // Reset password with token
  async resetPassword(request: ResetPasswordRequest): Promise<{ message: string }> {
    try {
      return await apiClient.post<{ message: string }>("/auth/reset-password", request)
    } catch (error) {
      throw error
    }
  }

  // Change password (authenticated)
  async changePassword(request: ChangePasswordRequest): Promise<{ message: string }> {
    try {
      return await apiClient.post<{ message: string }>("/auth/change-password", request)
    } catch (error) {
      throw error
    }
  }

  // Verify email with token
  async verifyEmail(token: string): Promise<{ message: string }> {
    try {
      return await apiClient.post<{ message: string }>("/auth/verify-email", { token })
    } catch (error) {
      throw error
    }
  }

  // Check if user is authenticated
  isAuthenticated(): boolean {
    return !!localStorage.getItem("token")
  }

  // Get current user from localStorage
  getUser(): User | null {
    const userStr = localStorage.getItem("user")
    if (!userStr) return null

    try {
      return JSON.parse(userStr) as User
    } catch (error) {
      console.error("Error parsing user data:", error)
      return null
    }
  }

  // Set auth session
  private setSession(token: string, refreshToken: string): void {
    localStorage.setItem("token", token)
    localStorage.setItem("refreshToken", refreshToken)
  }

  // Clear auth session
  private clearSession(): void {
    localStorage.removeItem("token")
    localStorage.removeItem("refreshToken")
    localStorage.removeItem("user")
  }
}

// Create and export a singleton instance
const authService = new AuthService()
export default authService

File: backend/src/main/java/com/quizmaster/service/AuthService.java
package com.quizmaster.service;

import com.quizmaster.dto.request.ForgotPasswordRequest;
import com.quizmaster.dto.request.LoginRequest;
import com.quizmaster.dto.request.RegisterRequest;
import com.quizmaster.dto.request.ResetPasswordRequest;
import com.quizmaster.dto.request.VerifyEmailRequest;
import com.quizmaster.dto.request.ChangePasswordRequest;
import com.quizmaster.dto.response.AuthResponse;
import com.quizmaster.dto.response.MessageResponse;
import com.quizmaster.model.User;
import com.quizmaster.repository.UserRepository;
import com.quizmaster.security.JwtService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;
    private final EmailService emailService;

    public AuthResponse register(RegisterRequest request) {
        // Check if username or email already exists
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new RuntimeException("Username is already taken");
        }

        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email is already in use");
        }

        // Create verification token
        String verificationToken = UUID.randomUUID().toString();
        LocalDateTime verificationTokenExpiry = LocalDateTime.now().plusDays(1);

        // Create user
        List<String> roles = new ArrayList<>();
        roles.add("USER");
        
        User user = User.builder()
                .username(request.getUsername())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .fullName(request.getFullName())
                .roles(roles)
                .emailVerified(false)
                .verificationToken(verificationToken)
                .verificationTokenExpiry(verificationTokenExpiry)
                .enabled(true)
                .locked(false)
                .build();

        userRepository.save(user);

        // Send verification email
        emailService.sendVerificationEmail(user.getEmail(), user.getFullName(), verificationToken);

        // Generate JWT token
        String jwtToken = jwtService.generateToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.builder()
                .token(jwtToken)
                .refreshToken(refreshToken)
                .build();
    }

    public AuthResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getUsername(),
                        request.getPassword()
                )
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);
        User user = (User) authentication.getPrincipal();

        String jwtToken = jwtService.generateToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.builder()
                .token(jwtToken)
                .refreshToken(refreshToken)
                .build();
    }

    public MessageResponse forgotPassword(ForgotPasswordRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("User not found"));

        String resetToken = UUID.randomUUID().toString();
        user.setResetPasswordToken(resetToken);
        user.setResetPasswordTokenExpiry(LocalDateTime.now().plusHours(1));

        userRepository.save(user);

        // Send reset password email
        emailService.sendResetPasswordEmail(user.getEmail(), user.getFullName(), resetToken);

        return new MessageResponse("Password reset email sent");
    }

    public MessageResponse resetPassword(ResetPasswordRequest request) {
        User user = userRepository.findByResetPasswordToken(request.getToken())
                .orElseThrow(() -> new RuntimeException("Invalid token"));

        // Check if token is expired
        if (user.getResetPasswordTokenExpiry().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Token expired");
        }

        // Update password
        user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        user.setResetPasswordToken(null);
        user.setResetPasswordTokenExpiry(null);

        userRepository.save(user);

        return new MessageResponse("Password reset successful");
    }

    public MessageResponse verifyEmail(VerifyEmailRequest request) {
        User user = userRepository.findByVerificationToken(request.getToken())
                .orElseThrow(() -> new RuntimeException("Invalid token"));

        // Check if token is expired
        if (user.getVerificationTokenExpiry().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Token expired");
        }

        // Update email verification status
        user.setEmailVerified(true);
        user.setVerificationToken(null);
        user.setVerificationTokenExpiry(null);

        userRepository.save(user);

        return new MessageResponse("Email verified successfully");
    }

    public MessageResponse changePassword(ChangePasswordRequest request) {
        // Get current authenticated user
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        User user = (User) authentication.getPrincipal();

        // Verify current password
        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPassword())) {
            throw new RuntimeException("Current password is incorrect");
        }

        // Update password
        user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(user);

        return new MessageResponse("Password changed successfully");
    }
}
